onnx/defs/experiments/defs.cc

```diff
@@ -55,13 +56,14 @@ NOTE: Currently, it supports data type of float, int32, int64, and bool.
         "1D tensor containing the desired output shape.  First input must be in "
         "CPU context.",
         AttrType::INT)
-    .Input(0, "input", "Input tensor (optional) to provide shape information.",
-           true /*optional*/)
+    .Input(0, "input", "Input tensor (optional) to provide shape information.", "T", true)
```

Comment:
This seems like an interesting example that may require further extensions to the type system to capture fully. We need to capture that "dtype" is the same as "T" (which we could do by using the same name for both). Something to think about as a future extension.